{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bf8a9252-0371-46eb-983b-177fee981685",
   "metadata": {},
   "source": [
    "# Project 03: Gibbs Sampling (Random Algorithm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "35409725-ccac-4013-bda5-ad8d03548608",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " UserWarning:/Users/marcosequiza/anaconda3/envs/project03/lib/python3.14/site-packages/weblogo/__init__.py:1: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n"
     ]
    }
   ],
   "source": [
    "import random as random\n",
    "import numpy as np\n",
    "import bamnostic as bs\n",
    "import seqlogo as sl\n",
    "\n",
    "#import function for building sequence motif & idenfitying seqs matching to motif\n",
    "from data_readers import get_fasta, get_gff\n",
    "from seq_ops import get_seq, reverse_complement\n",
    "from motif_ops import build_pfm, build_pwm, pfm_ic, score_kmer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a1ad95-1019-42fc-839c-0a457decbedf",
   "metadata": {},
   "source": [
    "---\n",
    "## Implement Gibbs Sampler\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51922616-8ff6-4d86-b3ef-ae714c03d1c4",
   "metadata": {},
   "source": [
    "Gibbs sampling is a MCMC approach to identify enrichments. Here we will implement a method to identify motifs from a set of regions. \n",
    "\n",
    "Important considerations:\n",
    "- We will need to score each sequence with a PWM using the `score_kmer()` or `score_sequence()` functions\n",
    "    - You will need to investigate into the help documenation and libraries to identify how best to use these functions. \n",
    "- These sites are often not strand-specific and so both scores on the negative as well as positive strand should be considered\n",
    "- To select a random sequence, use `random.randint()` or `numpy.random.randint()`\n",
    "- To select a new position $m$ (as defined below) use `random.choices()` or `numpy.random.choice()`  \n",
    "\n",
    "Assumptions: \n",
    "- We know $k$ as the length of expected motif\n",
    "- Each sequence contains the motif\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2601e5b1-6a3b-4f57-8eeb-e56f67cfda62",
   "metadata": {},
   "source": [
    "```\n",
    "GibbsMotifFinder(DNA, k-length)\n",
    "    random pick of k-length sequences from each line of DNA as Motifs\n",
    "    for j ← 1 to 10000 or Motifs stops changing\n",
    "        i ← Random(N) where N is number of DNA entries\n",
    "        PWM ← PWM constructed from all Motifs except for Motifi\n",
    "        Motifi ← select position m from PWM-scored k-mers in DNAi in probabilistic fashion from score distribution\n",
    "    return PFM\n",
    "```\n",
    "\n",
    "Probability of chosing position $m = \\frac{A_{m}}{\\sum_{l}A_{l}}$ for positions $l$ in DNAi\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bcc0c5a-f32d-4a3a-9b3f-fc33edaf2e7c",
   "metadata": {},
   "source": [
    "**Note:** I have also added a function to `motif_ops.py` that will calculate the information content of your motifs. This is useful to observe the progression of your Gibbs sampler as well as a measure of convergence. You can use this function as `IC = pfm_ic(pfm)`. You should expect a slow increase of IC until it plateaus such as in the plot below from your lecture slides:\n",
    "\n",
    "<center><img src='figures/Gibbs_Sampling.png'/ width=600px></center>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d218c76-f047-4e66-b1cc-3b1ffca510a0",
   "metadata": {},
   "source": [
    "---\n",
    "# Setting up the data\n",
    "\n",
    "This BAM file contains a subsampled set of aligned ChIP‑seq reads from a p53 immunoprecipitation experiment in human K562 cells treated with the anthracycline drug daunorubicin. The original SRA experiment SRX5865974 (run SRR9090854) reports 31.2 million Illumina NextSeq 500 ChIP‑seq read pairs from K562 wild‑type cells exposed to daunorubicin, using a p53 antibody to pull down p53‑bound chromatin fragments before library preparation and sequencing. [pmc.ncbi.nlm.nih](https://pmc.ncbi.nlm.nih.gov/articles/PMC4366240/)\n",
    "\n",
    "### What the data represent\n",
    "\n",
    "- Biological system: human K562 leukemia cells (Homo sapiens) treated with daunorubicin, a DNA‑damaging chemotherapeutic known to stabilize and activate p53. [pmc.ncbi.nlm.nih](https://pmc.ncbi.nlm.nih.gov/articles/PMC6561911/)\n",
    "- Assay: ChIP‑seq using an antibody against p53, so reads should be enriched around genomic regions where p53 is bound after drug treatment. [pmc.ncbi.nlm.nih](https://pmc.ncbi.nlm.nih.gov/articles/PMC4526040/)\n",
    "- Sequencing: Illumina NextSeq 500, with the raw run SRR9090854 corresponding to experiment SRX5865974. [github](https://github.com/ncbi/sra-tools/issues/213)\n",
    "- BAM file: `SRR9090854.subsampled_5pct.bam` is a coordinate‑sorted alignment file containing ~5% of the original mapped reads, typically created by random down‑sampling the full BAM to reduce file size and speed up exploratory analyses while preserving the overall distribution of p53 binding events. [ecseq](https://www.ecseq.com/support/ngs-snippets/how-to-extract-a-list-of-specific-read-IDs-from-a-BAM-file)\n",
    "\n",
    "### How `bamnostic` is used and what your code does\n",
    "\n",
    "The `bamnostic` package provides a pure‑Python interface to BAM files that mirrors the `pysam` API, including an `AlignmentFile` class whose iterator yields `AlignedSegment` objects representing individual aligned reads. In your code: [bamnostic.readthedocs](https://bamnostic.readthedocs.io/en/latest/bamnostic.html)\n",
    "\n",
    "```python\n",
    "bam_path = \"data/SRR9090854.subsampled_5pct.bam\"\n",
    "\n",
    "seqs = [read.seq for read in bs.AlignmentFile(bam_path)]\n",
    "```\n",
    "\n",
    "- `bs.AlignmentFile(bam_path)` opens the BAM file as an `AlignmentFile` object in binary read mode (default `'rb'`), reading the BAM header (reference contigs, read groups, etc.) and preparing a streaming interface to all aligned records. [github](https://github.com/betteridiot/bamnostic/blob/master/docs/source/quickstart.rst)\n",
    "- Iterating over `AlignmentFile` (`for read in bs.AlignmentFile(...)`) returns each aligned read as a `bamnostic.AlignedSegment` object, which exposes properties analogous to SAM fields such as query name, flags, reference name, position, mapping quality, CIGAR string, and the original sequencing **sequence**. [bamnostic.readthedocs](https://bamnostic.readthedocs.io/en/latest/bamnostic.html)\n",
    "- The `read.seq` attribute is the query (read) sequence string stored in the BAM, corresponding to the full read sequence (including any unaligned bases), as opposed to `query_alignment_sequence`, which would only contain the aligned portion. [bamnostic.readthedocs](https://bamnostic.readthedocs.io/en/latest/bamnostic.html)\n",
    "- The list comprehension `[read.seq for read in ...]` consumes the entire BAM stream and collects the nucleotide sequences from every subsampled ChIP‑seq read into a Python list `seqs`, which can then be used for downstream tasks such as motif discovery, k‑mer analysis, or quality checks on read content. [ucdavis-bioinformatics-training.github](https://ucdavis-bioinformatics-training.github.io/2022-Feb-Introduction-To-Python-For-Bioinformatics/python/python5)\n",
    "\n",
    "In summary, your dataset is a 5% random sample of p53‑ChIP‑seq alignments from daunorubicin‑treated K562 cells, and the `bamnostic` code opens the subsampled BAM and extracts the raw read sequences from each aligned fragment into memory as a list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "82427797-e842-40f8-929d-9d768e1498b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_path = \"data/SRR9090854.subsampled_5pct.bam\"\n",
    "\n",
    "seqs = [read.seq for read in bs.AlignmentFile(bam_path)] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "00f40be2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read promoters, store in list of strings\n",
    "seq_file = \"data/GCF_000009045.1_ASM904v1_genomic.fna\"\n",
    "gff_file = \"data/GCF_000009045.1_ASM904v1_genomic.gff\"\n",
    "\n",
    "seqss =[]\n",
    "for name, seq in get_fasta(seq_file):\n",
    "    for gff_entry in get_gff(gff_file):\n",
    "        if gff_entry.type == \"CDS\":\n",
    "            promoter_seq = get_seq(\n",
    "                seq,\n",
    "                gff_entry.start,\n",
    "                gff_entry.end,\n",
    "                gff_entry.strand,\n",
    "                50\n",
    "            )\n",
    "\n",
    "            if \"AGGAGG\" in promoter_seq:\n",
    "                seqss.append(promoter_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f117c460",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3083497, 837)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seqs), len(seqss)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f06d29d2-617a-4821-942c-b28975dc3e54",
   "metadata": {},
   "source": [
    "---\n",
    "# Project Start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "01de054f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def initialize_random_motif(seqs, k): \n",
    "    ''' Function to pick random motifs for each sequence '''\n",
    "\n",
    "    # Initialize motif list and iterate through each sequence\n",
    "    motif_sequence_list = []\n",
    "    for seq in seqs:\n",
    "\n",
    "        # Pick a random index for start of motif\n",
    "        motif_start_index = np.random.randint(low = 0, high = len(seq) - k + 1)\n",
    "\n",
    "        # Append motif to list based on index\n",
    "        motif_sequence_list.append(seq[motif_start_index:motif_start_index + k])\n",
    "        \n",
    "    return motif_sequence_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5ce422c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_new_motif(motif_list, seqs, k):\n",
    "    '''\n",
    "    Function iterates 1 time. Picks a sequence to omit, recalculates PWM, calculates kmer score for each possble motif in omitted seq, choses new motif stochastically\n",
    "    '''\n",
    "\n",
    "    # Randomly pick a sequence to omit and exclude from calculations\n",
    "    omitted_seq_ind = np.random.randint(low = 0, high = len(seqs))\n",
    "    omitted_seq = seqs[omitted_seq_ind]\n",
    "    temp_motif_list = motif_list[:omitted_seq_ind] + motif_list[omitted_seq_ind+1:]\n",
    "\n",
    "    # Recalculate PFM with temporary motif list\n",
    "    new_pfm = build_pfm(temp_motif_list, k)\n",
    "\n",
    "    # Recalculate PWM with temporary motif list\n",
    "    new_pwm = build_pwm(new_pfm)\n",
    "\n",
    "    # Create empty list to hold scores\n",
    "    scores = []\n",
    "\n",
    "    # Create sliding window to calculate score of every possible motif for forward (+) and reverse (-) strands\n",
    "    for i in range(len(omitted_seq) - k + 1):\n",
    "        motif = omitted_seq[i:i + k]\n",
    "        r_motif = reverse_complement(motif)\n",
    "\n",
    "        # Even indexes are (+) strand, odd indexes are (-)\n",
    "        scores.extend([score_kmer(motif, new_pwm), score_kmer(r_motif, new_pwm)])\n",
    "\n",
    "    #create sliding window to calculate score of every possible motif for forward strand    \n",
    "    #reverse_strand = reverse_complement(seqs[omitted_seq_ind])\n",
    "    #for i in range(len(reverse_strand) - k):\n",
    "        #r_motif = reverse_strand[i:i + k]\n",
    "        #scores.append(score_kmer(r_motif, new_pwm))\n",
    "    \n",
    "    # Normalize scores with softmax function (2^^score / sum(2^^score))\n",
    "    #normalize_scores = (2**scores)/sum(2**scores)\n",
    "    pow_scores = [(2**score) for score in scores]\n",
    "    total = sum(pow_scores)\n",
    "    normalize_scores = [s / total for s in pow_scores]\n",
    "\n",
    "    # Pick a \"random\" new motif index based on scores\n",
    "    new_motif_index = np.random.choice(len(normalize_scores), p = normalize_scores)\n",
    "\n",
    "    # Extract motif from (+) strand if it is an even index (divide by 2 to get actual index from sequence)\n",
    "    if new_motif_index % 2 == 0:\n",
    "        new_motif = omitted_seq[int(new_motif_index / 2) : int(new_motif_index / 2) + k]\n",
    "    \n",
    "    # Extract motif from (-) strand by reverse complement if it is an odd index (floor division by 2 to get actual index from sequence)\n",
    "    else:\n",
    "        new_motif = reverse_complement(omitted_seq[(new_motif_index // 2) : (new_motif_index // 2) + k])\n",
    "\n",
    "        # Update seqs list with reverse strand\n",
    "        seqs[omitted_seq_ind] = reverse_complement(omitted_seq)\n",
    "\n",
    "    #if new_motif_index > (len(normalize_scores)//2):\n",
    "        #forward_motif_index = new_motif_index - (len(normalize_scores)//2)\n",
    "        #new_motif = reverse_complement(seqs[omitted_seq_ind][forward_motif_index:forward_motif_index + k])\n",
    "        #seqs[omitted_seq_ind] = reverse_complement(seqs[omitted_seq_ind])\n",
    "    #else: \n",
    "        #new_motif = seqs[omitted_seq_ind][new_motif_index:new_motif_index + k]\n",
    "\n",
    "    return omitted_seq_ind, new_motif\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dccbe0a4-852c-4c15-a64b-893d00e7014f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def GibbsMotifFinder (seqs, k, seed=None, max_iterations=10000, ic_threshold=1e-8, convergence_iterations=300):\n",
    "\n",
    "    '''\n",
    "Function to find a pfm from a list of strings using a Gibbs sampler\n",
    "    \n",
    "    Args: \n",
    "        seqs (str list): a list of sequences, not necessarily in same lengths\n",
    "        k (int): the length of motif to find\n",
    "        seed (int, default=None): seed for np.random\n",
    "        max_iterations (int, default=10000): number of iterations to stop if there is no convergence\n",
    "        ic_threshold (float, default=1e-4): IC threshold to consider two values \"equal\"\n",
    "        convergence_iterations (int, default=100): number of iterations in a row to consider convergence reached \n",
    "\n",
    "    Returns:\n",
    "        pfm (numpy array): dimensions are 4xlength\n",
    "    '''\n",
    "    # Use rng to make random samples/selections/numbers\n",
    "    # Example: randint = rng.integer(1, 10)\n",
    "    random.seed(seed) \n",
    "    rng = np.random.default_rng(seed)\n",
    "    \n",
    "    # Select starting random motifs in each sequence\n",
    "    motif_list = initialize_random_motif(seqs, k)\n",
    "\n",
    "    # Create initial PFM\n",
    "    initial_pfm = build_pfm(motif_list, k)\n",
    "\n",
    "    # Create PWM from initial motif assignments\n",
    "    initial_pwm = build_pwm(initial_pfm)\n",
    "\n",
    "    # Initialize information_content for PFM\n",
    "    current_ic = pfm_ic(initial_pfm)\n",
    "\n",
    "    # Initialize iteration and convergence counters\n",
    "    i = 1\n",
    "    convergence_counter = 0\n",
    "\n",
    "    # Iterate until iteration limit\n",
    "    while i <= max_iterations:\n",
    "\n",
    "        # Pick a new motif for a randomly selected sequence \n",
    "        seq_ind, new_motif = select_new_motif(motif_list, seqs, k)\n",
    "\n",
    "        # Update motif list with newly picked motif\n",
    "        motif_list[seq_ind] = new_motif\n",
    "\n",
    "        # Calculate information content for new PFM\n",
    "        new_pfm = build_pfm(motif_list, k)\n",
    "        new_ic = pfm_ic(new_pfm)\n",
    "\n",
    "        # Calculate difference between old and current ICs to evaluate convergence\n",
    "        inf_content_delta = abs(new_ic - current_ic)\n",
    "\n",
    "        # Add +1 to convergence counter if IC difference is smaller than IC threshold\n",
    "        if inf_content_delta < ic_threshold:\n",
    "            convergence_counter += 1\n",
    "\n",
    "        # Restart covergence counter if IC difference is greater than threshold\n",
    "        else:\n",
    "            convergence_counter = 0\n",
    "\n",
    "        # If convergence is reached, return PFM \n",
    "        if convergence_counter == convergence_iterations: \n",
    "            return new_pfm\n",
    "        \n",
    "        # Update current IC and iteration counter for next iteration\n",
    "        current_ic = new_ic\n",
    "\n",
    "        if i % 10 == 0:\n",
    "            print(\"Iteration:\", i)\n",
    "            print(\"IC = \", current_ic)\n",
    "        i += 1\n",
    "    \n",
    "    # If loop ends without convergence being reached, output message and return latest PFM\n",
    "    if i >= max_iterations:\n",
    "        print(\"failed to converge...\")\n",
    "        return new_pfm\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab1f791-6765-46fd-a695-2c6a78430a3c",
   "metadata": {},
   "source": [
    "---\n",
    "# Driver Program\n",
    "Don't change any of the code here. If you have completed the project by following the coding by contract, the following code should work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0a730251-97cb-42df-aea9-eaccd0d03131",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration: 10\n",
      "IC =  0.0836546339553359\n",
      "Iteration: 20\n",
      "IC =  0.08365363154744476\n",
      "Iteration: 30\n",
      "IC =  0.08365357729891976\n",
      "Iteration: 40\n",
      "IC =  0.08365433312564763\n",
      "Iteration: 50\n",
      "IC =  0.08365418852608952\n",
      "Iteration: 60\n",
      "IC =  0.08365446326377635\n",
      "Iteration: 70\n",
      "IC =  0.08365446457325554\n",
      "Iteration: 80\n",
      "IC =  0.08365414208565047\n",
      "Iteration: 90\n",
      "IC =  0.08365428106867578\n",
      "Iteration: 100\n",
      "IC =  0.08365477488865891\n",
      "Iteration: 110\n",
      "IC =  0.08365500098243839\n",
      "Iteration: 120\n",
      "IC =  0.08365527333917622\n",
      "Iteration: 130\n",
      "IC =  0.0836550056149663\n",
      "Iteration: 140\n",
      "IC =  0.08365650139590874\n",
      "Iteration: 150\n",
      "IC =  0.08365669935745901\n",
      "Iteration: 160\n",
      "IC =  0.08365707372862796\n",
      "Iteration: 170\n",
      "IC =  0.08365765446894469\n",
      "Iteration: 180\n",
      "IC =  0.08365761215575418\n",
      "Iteration: 190\n",
      "IC =  0.08365788313480227\n",
      "Iteration: 200\n",
      "IC =  0.08365944151007865\n",
      "Iteration: 210\n",
      "IC =  0.08365888981669345\n",
      "Iteration: 220\n",
      "IC =  0.08365916860395406\n",
      "Iteration: 230\n",
      "IC =  0.08366007692812905\n",
      "Iteration: 240\n",
      "IC =  0.08366093375096018\n",
      "Iteration: 250\n",
      "IC =  0.08366067441935021\n",
      "Iteration: 260\n",
      "IC =  0.08366050016624249\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[8]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# Run the gibbs sampler:\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m promoter_pfm = \u001b[43mGibbsMotifFinder\u001b[49m\u001b[43m(\u001b[49m\u001b[43mseqs\u001b[49m\u001b[43m,\u001b[49m\u001b[32;43m10\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m      4\u001b[39m \u001b[38;5;66;03m# Plot the final pfm that is generated: \u001b[39;00m\n\u001b[32m      5\u001b[39m sl.seqlogo(sl.CompletePm(pfm = promoter_pfm.T))\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[7]\u001b[39m\u001b[32m, line 42\u001b[39m, in \u001b[36mGibbsMotifFinder\u001b[39m\u001b[34m(seqs, k, seed, max_iterations, ic_threshold, convergence_iterations)\u001b[39m\n\u001b[32m     38\u001b[39m \u001b[38;5;66;03m# Iterate until iteration limit\u001b[39;00m\n\u001b[32m     39\u001b[39m \u001b[38;5;28;01mwhile\u001b[39;00m i <= max_iterations:\n\u001b[32m     40\u001b[39m \n\u001b[32m     41\u001b[39m     \u001b[38;5;66;03m# Pick a new motif for a randomly selected sequence \u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m42\u001b[39m     seq_ind, new_motif = \u001b[43mselect_new_motif\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmotif_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mseqs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     44\u001b[39m     \u001b[38;5;66;03m# Update motif list with newly picked motif\u001b[39;00m\n\u001b[32m     45\u001b[39m     motif_list[seq_ind] = new_motif\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[6]\u001b[39m\u001b[32m, line 12\u001b[39m, in \u001b[36mselect_new_motif\u001b[39m\u001b[34m(motif_list, seqs, k)\u001b[39m\n\u001b[32m      9\u001b[39m temp_motif_list = motif_list[:omitted_seq_ind] + motif_list[omitted_seq_ind+\u001b[32m1\u001b[39m:]\n\u001b[32m     11\u001b[39m \u001b[38;5;66;03m# Recalculate PFM with temporary motif list\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m12\u001b[39m new_pfm = \u001b[43mbuild_pfm\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtemp_motif_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     14\u001b[39m \u001b[38;5;66;03m# Recalculate PWM with temporary motif list\u001b[39;00m\n\u001b[32m     15\u001b[39m new_pwm = build_pwm(new_pfm)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/NU/classes/BINF6250_spencer/projects/project03/motif_ops.py:28\u001b[39m, in \u001b[36mbuild_pfm\u001b[39m\u001b[34m(sequences, length)\u001b[39m\n\u001b[32m     25\u001b[39m base_to_index[\u001b[38;5;28mord\u001b[39m(\u001b[33m'\u001b[39m\u001b[33mT\u001b[39m\u001b[33m'\u001b[39m)] = \u001b[32m3\u001b[39m\n\u001b[32m     27\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m seq \u001b[38;5;129;01min\u001b[39;00m sequences:\n\u001b[32m---> \u001b[39m\u001b[32m28\u001b[39m     seq_array = np.frombuffer(\u001b[43mseq\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43mlength\u001b[49m\u001b[43m]\u001b[49m\u001b[43m.\u001b[49m\u001b[43mencode\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m, dtype=np.int8)\n\u001b[32m     29\u001b[39m     indices = base_to_index[seq_array]\n\u001b[32m     30\u001b[39m     np.add.at(pfm, (indices, np.arange(length)), \u001b[32m1\u001b[39m)\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "# Run the gibbs sampler:\n",
    "promoter_pfm = GibbsMotifFinder(seqs,10 )\n",
    "\n",
    "# Plot the final pfm that is generated: \n",
    "sl.seqlogo(sl.CompletePm(pfm = promoter_pfm.T))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a22848f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "promoter_pfm"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BINF6250_class-S1fKS3PI",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
